src/gherkin/testContext.test.js

```diff
@@ -0,0 +1,261 @@
+      return 0;
+    });
+    try {
+      testContext.addstep("GIVEN", "user {string} has been created", () => {
+        return 0;
+      });
+      fail();
+    } catch (err) {
+      expect(err.message).toBe("step already registered");
+    }
+  });
+
+  it("test given when and then functions", () => {
+    const testContext = new TestContext();
+
+    testContext.given("user {string} has been created", actionFunction);
+    testContext.given(
+      "user {string} has been created with default attributes",
+      actionFunction
+    );
+    testContext.when("admin creates a user {string}", actionFunction);
+    testContext.when(
+      "admin creates a user {string} with default attributes",
+      actionFunction
+    );
+    testContext.then("user {string} should be created", actionFunction);
+    testContext.then(
+      "user {string} should be created with default attributes",
+      actionFunction
+    );
+
+    const expectedSteps = [
+      new StepDef("GIVEN", "user {string} has been created", actionFunction),
+      new StepDef(
+        "GIVEN",
+        "user {string} has been created with default attributes",
+        actionFunction
+      ),
+      new StepDef("WHEN", "admin creates a user {string}", actionFunction),
+      new StepDef(
+        "WHEN",
+        "admin creates a user {string} with default attributes",
+        actionFunction
+      ),
+      new StepDef("THEN", "user {string} should be created", actionFunction),
+      new StepDef(
+        "THEN",
+        "user {string} should be created with default attributes",
+        actionFunction
+      ),
+    ];
+    for (let i = 0; i < expectedSteps.length; i++) {
+      expect(testContext.steps[i].token).toStrictEqual(expectedSteps[i].token);
+      expect(testContext.steps[i].pattern).toStrictEqual(
+        expectedSteps[i].pattern
+      );
+      expect(testContext.steps[i].action).toStrictEqual(
+        expectedSteps[i].action
+      );
+    }
+  });
+
+  it("test getMatch", () => {
+    const testContext = new TestContext();
+    const expectedStep = [];
+
+    for (const Data of getMatchData) {
+      let step;
+      testContext.addstep(
+        Data.addStep.token,
+        Data.addStep.step,
+        Data.addStep.action
+      );
+
+      if (Data.newStep.table) {
+        step = new Step(
+          Data.newStep.token,
+          Data.newStep.step,
+          Data.newStep.table
+        );
+      } else {
+        step = new Step(Data.newStep.token, Data.newStep.step);
+      }
+
+      expectedStep.push(testContext.getMatch(step));
+    }
+    expect(testContext.steps).toStrictEqual(expectedStep);
+  });
+
+  it("test before and after function", () => {
+    const testContext = new TestContext();
+    testContext.before("abc");
+    expect(testContext.beforeSteps).toStrictEqual(["abc"]);
+    testContext.after("xyz");
+    expect(testContext.afterSteps).toStrictEqual(["xyz"]);
+  });
```

Comment:
before and after should only expect functions, you should check that also test calling setup() and cleanup()

Reply:
fixed